js.UndefinedParam interacts badly with compiler plugins

I have a cross project and when using the "test" command in SBT, I get the following error:

```
org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$OptimizeException: The Scala.js optimizer crashed while optimizing Lsome_
package_AbstractAjaxLoader.init___T__T: java.lang.RuntimeException: Invalid tree in transform of class org.scalajs.core.ir.Trees$UndefinedParam: UndefinedParam()
Methods attempted to inline:
* O.init___
* s_concurrent_Promise$.apply__s_concurrent_Promise
* s_concurrent_impl_Promise$DefaultPromise.future__s_concurrent_Future
* s_concurrent_impl_Promise$DefaultPromise.future__s_concurrent_impl_Promise
[error] * static s_concurrent_impl_Promise$class.future__s_concurrent_impl_Promise__s_concurrent_impl_Promise
```

The strange thing is that this only happens with the "test" command. fastOptJS works perfectly !

Here's the full class where the error seems to happen:

```
abstract class AbstractAjaxLoader(val path: String, val responseType: String) extends Loader {

    type Data = dom.XMLHttpRequest

    protected override val data: Future[Data] = {
        val promise = Promise[Data]()

        val request = new XMLHttpRequest()
        request.open(method = "GET", url = path)
        request.responseType = responseType
        request.onload = (_: dom.Event) => promise.success(request)
        request.onerror =
                (e: ErrorEvent) => promise.failure(new Exception(s"Failed to load $path: ${e.message}"))

        request.send()
        promise.future
    }

    def dataText: Future[String] = data.map(_.responseText)

}
```

And the full stack:

```
[info] Fast optimizing /.../js/target/scala-2.11/project-test-fastopt.js
org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$OptimizeException: The Scala.js optimizer crashed while optimizing Lsome_package_AbstractAjaxLoader.init___T__T: java.lang.RuntimeException: Invalid tree in transform of class org.scalajs.core.ir.Trees$UndefinedParam: UndefinedParam()
Methods attempted to inline:
* O.init___
* s_concurrent_Promise$.apply__s_concurrent_Promise
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:150)
    at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.process(GenIncOptimizer.scala:921)
    at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:87)
    at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:86)
    at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach_quick(ParArray.scala:145)
    at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach(ParArray.scala:138)
    at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:975)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
    at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
    at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:972)
    at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
    at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
    at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
    at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
    at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.RuntimeException: Invalid tree in transform of class org.scalajs.core.ir.Trees$UndefinedParam: UndefinedParam()
    at scala.sys.package$.error(package.scala:27)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:677)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:701)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:722)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:313)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformExpr(OptimizerCore.scala:283)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformExprOrSpread(OptimizerCore.scala:291)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$34.apply(OptimizerCore.scala:626)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$34.apply(OptimizerCore.scala:626)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:244)
    at scala.collection.immutable.List.foreach(List.scala:318)
    at scala.collection.TraversableLike$class.map(TraversableLike.scala:244)
    at scala.collection.AbstractTraversable.map(Traversable.scala:105)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:626)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformStat(OptimizerCore.scala:271)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:715)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1$1$$anonfun$apply$9$$anonfun$apply$10.apply(OptimizerCore.scala:708)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1$1$$anonfun$apply$9$$anonfun$apply$10.apply(OptimizerCore.scala:707)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBinding$1.apply(OptimizerCore.scala:3774)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBinding$1.apply(OptimizerCore.scala:3773)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.withDedicatedVar$1(OptimizerCore.scala:3850)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.apply(OptimizerCore.scala:3882)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.apply(OptimizerCore.scala:3806)
    at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
    at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:3975)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:704)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1$1$$anonfun$apply$9$$anonfun$apply$10.apply(OptimizerCore.scala:708)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1$1$$anonfun$apply$9$$anonfun$apply$10.apply(OptimizerCore.scala:707)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBinding$1.apply(OptimizerCore.scala:3774)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withBinding$1.apply(OptimizerCore.scala:3773)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.withDedicatedVar$1(OptimizerCore.scala:3850)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.apply(OptimizerCore.scala:3882)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$withNewLocalDef$1.apply(OptimizerCore.scala:3806)
    at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
    at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:3975)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:704)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:722)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:313)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformExpr(OptimizerCore.scala:283)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$9.apply(OptimizerCore.scala:357)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$9.apply(OptimizerCore.scala:322)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon(OptimizerCore.scala:1070)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1003)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformSelectCommon$1.apply(OptimizerCore.scala:1001)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:793)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore$$anonfun$org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$pretransformExpr$1.apply(OptimizerCore.scala:774)
    at scala.util.control.TailCalls$TailRec.loop$1(TailCalls.scala:37)
    at scala.util.control.TailCalls$TailRec.result(TailCalls.scala:40)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$trampoline(OptimizerCore.scala:3975)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:360)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:701)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transformList$1(OptimizerCore.scala:717)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformBlock(OptimizerCore.scala:722)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.org$scalajs$core$tools$linker$frontend$optimizer$OptimizerCore$$transform(OptimizerCore.scala:313)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.transformIsolatedBody(OptimizerCore.scala:3617)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.liftedTree1$1(OptimizerCore.scala:131)
    at org.scalajs.core.tools.linker.frontend.optimizer.OptimizerCore.optimize(OptimizerCore.scala:130)
    at org.scalajs.core.tools.linker.frontend.optimizer.GenIncOptimizer$MethodImpl.process(GenIncOptimizer.scala:921)
    at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:87)
    at org.scalajs.core.tools.linker.frontend.optimizer.ParIncOptimizer$$anonfun$processAllTaggedMethods$2.apply(ParIncOptimizer.scala:86)
    at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach_quick(ParArray.scala:145)
    at scala.collection.parallel.mutable.ParArray$ParArrayIterator.foreach(ParArray.scala:138)
    at scala.collection.parallel.ParIterableLike$Foreach.leaf(ParIterableLike.scala:975)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply$mcV$sp(Tasks.scala:54)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
    at scala.collection.parallel.Task$$anonfun$tryLeaf$1.apply(Tasks.scala:53)
    at scala.collection.parallel.Task$class.tryLeaf(Tasks.scala:56)
    at scala.collection.parallel.ParIterableLike$Foreach.tryLeaf(ParIterableLike.scala:972)
    at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.internal(Tasks.scala:183)
    at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.internal(Tasks.scala:514)
    at scala.collection.parallel.AdaptiveWorkStealingTasks$WrappedTask$class.compute(Tasks.scala:162)
    at scala.collection.parallel.AdaptiveWorkStealingForkJoinTasks$WrappedTask.compute(Tasks.scala:514)
    at scala.concurrent.forkjoin.RecursiveAction.exec(RecursiveAction.java:160)
    at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
    at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
    at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
    at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

--- a/compiler/src/main/scala/org/scalajs/core/compiler/GenJSFiles.scala
+++ b/compiler/src/main/scala/org/scalajs/core/compiler/GenJSFiles.scala
@@ -29,6 +29,20 @@ trait GenJSFiles extends SubComponent { self: GenJSCode =>
     try {
       ir.InfoSerializers.serialize(output, Infos.generateClassInfo(tree))
       ir.Serializers.serialize(output, tree)
+    } catch {
+      case e: ir.InvalidIRException =>
+        e.tree match {
+          case ir.Trees.UndefinedParam() =>
+            reporter.error(sym.pos, "Found a dangling UndefinedParam at " +
+                s"${e.tree.pos}. This is likely due to a bad interaction " +
+                "between a macro or a compiler plugin and the Scala.js " +
+                "compiler plugin. If you hit this, please let us know.")
+
+          case _ =>
+            reporter.error(sym.pos, "The Scala.js compiler generated " +
+                "invalid IR for this class. Please report this as a bug. IR: " +
+                e.tree)
+        }
     } finally {
       output.close()
     }

--- a/compiler/src/test/scala/org/scalajs/core/compiler/test/JSUndefinedParamTest.scala
+++ b/compiler/src/test/scala/org/scalajs/core/compiler/test/JSUndefinedParamTest.scala
@@ -1,0 +1,89 @@
+package org.scalajs.core.compiler.test
+
+import org.scalajs.core.compiler.test.util._
+import org.junit.Test
+
+// scalastyle:off line.size.limit
+
+/** This tests the UndefinedParam tracker in the compiler backend.
+ *
+ *  In order to properly implement removal of trailing default parameters, the
+ *  compiler backend may generate a UndefinedParam tree and catch it later.
+ *  However, some macros and compiler plugins may generate trees the backend
+ *  doesn't expect. As a result the backend used to generate invalid IR.
+ *  Instead, we now track these helper trees and emit a more helpful error
+ *  message if one of them sticks around.
+ *
+ *  This test contains a macro that generates a tree we cannot handle and
+ *  verifies that the compiler bails out.
+ */
+class JSUndefinedParamTest extends DirectTest with TestHelpers {
+
+  /* We need a macro in the test. Therefore, we add scala-reflect and the
+   * compiler's output path itself to the classpath.
+   */
+  override def classpath: List[String] =
+    super.classpath ++ List(scalaReflectPath, testOutputPath)
+
+  @Test def noDanglingUndefinedParam: Unit = {
+
+    // Define macro that extracts method parameter.
+    """
+    import language.experimental.macros
+
+    /** Dummy object to get the right shadowing for cross compilation */
+    private object Compat210 {
+      object blackbox { // scalastyle:ignore
+        type Context = scala.reflect.macros.Context
+      }
+    }
+
+    import Compat210._
+
+    object JSUndefinedParamTest {
+      import scala.reflect.macros._ // shadows blackbox from above
+      import blackbox.Context
+
+      def extractArg(call: Any): Any = macro extractArg_impl
+
+      def extractArg_impl(c: Context)(call: c.Expr[Any]): c.Expr[Any] = {
+        import c.universe._
+
+        call.tree match {
+          case Apply(fun, List(arg)) => c.Expr[Any](arg)
+
+          case tree =>
+            c.abort(tree.pos, "Bad tree. Need function call with single argument.")
+        }
+      }
+    }
+    """.succeeds()
+
+    // Use the macro to trigger UndefinedParam catcher.
+    """
+    import scala.scalajs.js
+    import scala.scalajs.js.annotation._
+
+    @js.native
+    trait MyTrait extends js.Any {
+      def foo(x: Int = js.native): Int = js.native
+    }
+
+    object A {
+      val myTrait: MyTrait = ???
+
+      /* We assign the default parameter value for foo to b.
+       * This should fail.
+       */
+      val b = JSUndefinedParamTest.extractArg(myTrait.foo())
+    }
+    """ hasErrors
+    """
+      |newSource1.scala:10: error: Found a dangling UndefinedParam at Position(virtualfile:newSource1.scala,15,54). This is likely due to a bad interaction between a macro or a compiler plugin and the Scala.js compiler plugin. If you hit this, please let us know.
+      |    object A {
+      |           ^
+    """
+
+  }
+
+}

--- a/compiler/src/test/scala/org/scalajs/core/compiler/test/util/DirectTest.scala
+++ b/compiler/src/test/scala/org/scalajs/core/compiler/test/util/DirectTest.scala
@@ -8,6 +8,8 @@ import org.scalajs.core.compiler.ScalaJSPlugin
 
 import scala.collection.mutable
 
+import java.io.File
+
 /** This is heavily inspired by scala's partest suite's DirectTest */
 abstract class DirectTest {
 
@@ -26,7 +28,7 @@ abstract class DirectTest {
         List(
             "-d", testOutputPath,
             "-bootclasspath", scalaLibPath,
-            "-classpath", scalaJSLibPath) ++
+            "-classpath", classpath.mkString(File.pathSeparator)) ++
         extraArgs ++ args.toList)
 
     lazy val global: Global = new Global(settings, newReporter(settings)) {
@@ -68,8 +70,16 @@ abstract class DirectTest {
   // Filed as #1443
   def defaultGlobal: Global = newScalaJSCompiler()
 
-  def testOutputPath: String = sys.props("scala.scalajs.compiler.test.output")
+  def testOutputPath: String = {
+    val baseDir = sys.props("scala.scalajs.compiler.test.output")
+    val outDir = new File(baseDir, getClass.getName)
+    outDir.mkdirs()
+    outDir.getAbsolutePath
+  }
+
   def scalaJSLibPath: String = sys.props("scala.scalajs.compiler.test.scalajslib")
   def scalaLibPath: String   = sys.props("scala.scalajs.compiler.test.scalalib")
+  def scalaReflectPath: String = sys.props("scala.scalajs.compiler.test.scalareflect")
 
+  def classpath: List[String] = List(scalaJSLibPath)
 }

--- a/ir/src/main/scala/org/scalajs/core/ir/Infos.scala
+++ b/ir/src/main/scala/org/scalajs/core/ir/Infos.scala
@@ -382,6 +382,14 @@ object Infos {
         case LoadJSModule(ClassType(cls)) =>
           builder.addAccessedModule(cls)
 
+        /* We explicitly catch UndefinedParam here to make sure, we do not
+         * emit it in the compiler. This does not entirely belong here, as it
+         * supports GenJSCode, but it is not wrong to throw an exception.
+         */
+        case UndefinedParam() =>
+          throw new InvalidIRException(tree,
+              "Found UndefinedParam while building infos")
+
         case _ =>
       }
 

--- a/ir/src/main/scala/org/scalajs/core/ir/InvalidIRException.scala
+++ b/ir/src/main/scala/org/scalajs/core/ir/InvalidIRException.scala
@@ -1,0 +1,4 @@
+package org.scalajs.core.ir
+
+class InvalidIRException(val tree: Trees.Tree, message: String)
+    extends Exception(message)

--- a/project/Build.scala
+++ b/project/Build.scala
@@ -529,21 +529,29 @@ object Build {
               testOutDir.getAbsolutePath
             sys.props("scala.scalajs.compiler.test.scalajslib") =
               (packageBin in (library, Compile)).value.getAbsolutePath
-            sys.props("scala.scalajs.compiler.test.scalalib") = {
 
-              def isScalaLib(att: Attributed[File]) = {
+            def scalaArtifact(name: String): String = {
+              def isTarget(att: Attributed[File]) = {
                 att.metadata.get(moduleID.key).exists { mId =>
                   mId.organization == "org.scala-lang" &&
-                  mId.name         == "scala-library"  &&
-                  mId.revision     == scalaVersion.value
+                  mId.name == name &&
+                  mId.revision == scalaVersion.value
                 }
               }
 
-              val lib = (managedClasspath in Test).value.find(isScalaLib)
-              lib.map(_.data.getAbsolutePath).getOrElse {
-                streams.value.log.error("Couldn't find Scala library on the classpath. CP: " + (managedClasspath in Test).value); ""
+              (managedClasspath in Test).value.find(isTarget).fold {
+                streams.value.log.error(s"Couldn't find $name on the classpath")
+                ""
+              } { lib =>
+                lib.data.getAbsolutePath
               }
             }
+
+            sys.props("scala.scalajs.compiler.test.scalalib") =
+              scalaArtifact("scala-library")
+
+            sys.props("scala.scalajs.compiler.test.scalareflect") =
+              scalaArtifact("scala-reflect")
           },
           exportJars := true
       )

